toolbox/fdc3-for-web/fdc3-web-impl/src/FDC3ServerInstance.ts

```diff
@@ -0,0 +1,312 @@
+import { Directory } from './directory/DirectoryInterface';
+
+export enum HeartbeatActivityEvent {
+  ConnectedResponding = 'ConnectedResponding',
+  NotRespondingAfterDisconnectTime = 'NotRespondingAfterDisconnectTime',
+  NotRespondingAfterDeadTime = 'NotRespondingAfterDeadTime',
+}
+
+export type ContextListenerRegistration = {
+  appId: string;
+  instanceId: string;
+  listenerUuid: string;
+  channelId: string | null;
+  contextType: string | null;
+};
+
+export type PrivateChannelEventListener = {
+  appId: string;
+  instanceId: string;
+  channelId: string;
+  eventType: PrivateChannelEventTypes | null;
+  listenerUuid: string;
+};
+
+export type DesktopAgentEventListener = {
+  appId: string;
+  instanceId: string;
+  eventType: string | null;
+  listenerUuid: string;
+};
+
+export enum ChannelType {
+  'user',
+  'app',
+  'private',
+}
+
+export type ChannelState = {
+  id: string;
+  type: ChannelType;
+  context: Context[];
+  displayMetadata: DisplayMetadata;
+};
+
+export type IntentListenerRegistration = {
+  appId: string;
+  instanceId: string;
+  intentName: string;
+  listenerUUID: string;
+};
+
+/**
+ * Handles messaging to apps and opening apps for ONE FDC3 environment.
+ * Stores all state for MessageHandlers to use.
+ */
+export interface FDC3ServerInstance {
+  /**
+   * Returns the directory used by this FDC3ServerInstance.
+   */
+  getDirectory(): Directory;
+
+  /**
+   * UUID for outgoing message
+   */
+  createUUID(): string;
+
+  /**
+   * Post an outgoing message to a particular app
+   */
+  post(message: object, instanceId: InstanceID): Promise<void>;
+
+  /**
+   * Opens a new instance of an application.
+   * Promise completes once the application window is opened
+   */
+  open(appId: string): Promise<InstanceID>;
+
+  /**
+   * Registers a particular instance id with a given app id
+   */
+  setInstanceDetails(uuid: InstanceID, details: AppRegistration): void;
+
+  /**
+   * Returns the connection details for a particular instance of an app.
+   * Used in a variety of MessageHandler classes to retrieve details for
+   * an app and when validating an app's identity when connecting.
+   */
+  getInstanceDetails(uuid: InstanceID): AppRegistration | undefined;
+
+  /**
+   * Registers an app as connected to the desktop agent.
+   */
+  setAppState(app: InstanceID, state: State): Promise<void>;
+
+  /**
+   * Handle heartbeat activity from the HeartbeatHandler.
+   * The server instance can decide how to respond to different heartbeat events.
+   */
+  heartbeatActivity(instanceId: InstanceID, event: HeartbeatActivityEvent): Promise<void>;
+
+  /**
+   * Returns the list of apps open and connected to FDC3 at the current time.
+   * Note, it is the implementor's job to ensure this list is
+   * up-to-date in the event of app crashes or disconnections.
+   */
+  getConnectedApps(): Promise<AppRegistration[]>;
+
+  /**
+   * Return the list of all apps that have ever been registered with the ServerContext.
```

Comment:
I'd also consider updating this comment to remove 'ever been' as a long running implementation might drop older registrations, or ones it knows are gone forever (say a window closed so there no way its coming back as the same instance ever again - the instance details in SessionStorage dies with the window). Also clarify that its instances/registrations rather than app definitions.
```suggestion
   * Return the list of all app instances that have been registered with the ServerContext.
```